# 2023년 10월 29일(일)

---

### Today I Learned 

- [iOS-Swift] 에러처리 (Review)

---

## 에러 종류

1. 컴파일 타임 에러
   - Swift 문법과 관련된 에러(컴파일러가 미리 알고 수정해야한다고 알려줌)
2. 런타임 에러
   - 프로그램이 실행되는 동안 발생
   - 크래시(앱이 강제로 종료)
   - 발생 가능한 에러를 미리 처리해 두면, 강제 종료되지 않음 (개발자가 처리해야만 하는 에러)

## (런타임) 에러 처리가 왜 필요할까?

- 앱이 실행하는 중간에 왜 꺼질까?
- 어떤 예기치 못한 상황이 발생할 수 있음
  - (네트워크 통신을 하는 등의 경우에서) 서버에서 데이터를 못 가져와서 --> 꺼짐
- 앱이 그냥 꺼지는 게 좋은가? 아니면 예를들어 "서버에서 문제가 발생했습니다. 잠시 뒤에 다시 접속해 주세요"라고 알려주는 것이 좋은가?
- 프로세스 중에서, 예외적인 상황(에러)이 발생하는 것이 미리 가능성 등을 처리해 놓으면 앱이 무작정 꺼지는 것을 예방할 수 있음⭐️
- 에러는 일반적으로 동작, 즉, 함수의 처리과정에서 일어남
- 함수를 정의할 때, 예외적인 일이 발생하는 경우가 발생할 수 있는 함수이다. 라고 정의하고 처리하는 과정을 배울 것임.
- (에러 발생 가능) 함수 ---> 함수 실행시에 조금 다르게 처리해야 함(에러 처리)⭐️

## 에러 처리의 과정 (3단계)

```swift
// 1) 에러 정의 (어떤 에러가 발생할지 경우를 미리 정의)
enum HeightError: Error { // 에러 프로토콜 채택 (약속)
    case maxHeight
    case minHeight
}

// 2) 에러가 발생할 수 있는 함수에 대한 정의
// 에러 발생가능한 상황에 throw 키워드로 에러를 던질(throwing)수 있도록 정의 
func checkingHeight(height: Int) throws -> Bool { // (에러를 던질 수 있는 함수 타입이다.)
    if height > 190 {
        throw HeightError.maxHeight
    } else if height < 130 {
        throw HeightError.minHeight
    } else {
        if height >= 160 {
            return true
        } else {
            return false
        }
    }
}

// 3) 에러가 발생할 수 있는 함수의 처리(함수의 실행) ---> 기존과 조금 다르다. (try와 do-catch 문으로 처리)
// 정식적인 처리방법: do-catch 블럭에서 처리(조금 간편한 방식도 존재)
do {
    let isChecked = try checkingHeight(height: 200)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch {
    print("놀이기구 타는 것 불가능")
}
```

- do 블럭 - 함수를 통한 정상적인 처리의 경우 실행하는 블럭
- catch 블럭 - 함수가 에러를 던졌을 경우의 처리 실행하는 블럭

## 에러를 처리하는 방법 - try / try? / try!

```swift
/**===========================================
 1) 에러 정식 처리 방법
- 모든 에러발생의 예외적인 경우를 디테일하게 처리 가능
============================================**/
do {
    let isChecked = try checkingHeight(height: 200)
  
    if isChecked {
        print("청룡열차 가능")
    } else {
        print("후룸라이드 가능")
    }
} catch {
    print("놀이기구 타는 것 불가능")
}

/**===========================================
 2) try? (옵셔널 트라이) ===> 옵셔널 타입으로 리턴

   (1) 정상적인 경우 ==> (정상)리턴타입으로 리턴
   (2) 에러가 발생하면 ==> nil 리턴
============================================**/
let isChecked = try? checkingHeight(height: 200)      // Bool? --> 당연히, 옵셔널 타입을 벗겨서 사용해야함

/**===========================================
 3) try! (Forced 트라이) ===> 에러가 날 수 없는 경우에만 사용 가능

   (1) 정상적인 경우 ==> (정상)리턴타입으로 리턴
   (2) 에러가 발생하면 ==> 런타임에러
============================================**/
let isChecked2: Bool = try! checkingHeight(height: 150)      // Bool --> 에러가 발생할 수 없다고 확신이 있는 경우만 사용해야 함.
```

## Catch블럭 처리법

```swift
// catch블럭은 do블럭에서 발생한 에러만을 처리하는 블럭
// 모든 에러를 반드시 처리해야만 함 (글로벌 스코프에서는 모든 에러를 처리하지 않아도 컴파일 에러발생하지 않음)

// 패턴이 있는 경우(모든 에러를 각각 따로 처리 해야함)
do {
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch HeightError.maxHeight  {    // where절을 추가해서, 매칭시킬 에러패턴에 조건을 추가할 수 있음
    print("키가 커서 놀이기구 타는 것 불가능")
} catch HeightError.minHeight {      // 생략가능
    print("키가 작아서 놀이기구 타는 것 불가능")
}

// catch 패턴이 없이 처리도 가능
do {
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 타는 것 가능: \(isChecked)")
} catch {    // error 상수를 제공 (모든 에러가 넘어옴)
    print(error.localizedDescription)
 
    if let error = error as? HeightError {    // 실제 우리가 정의한 구체적인 에러 타입이 아니고, 에러 타입(프로토콜)이 넘어올 뿐
        switch error {
        case .maxHeight:
            print("키가 커서 놀이기구 타는 것 불가능")
        case .minHeight:
            print("키가 작아서 놀이기구 타는 것 불가능")
        }
    }
}

// Swift 5.3이후 catch블럭의 처리 
do {   
    let isChecked = try checkingHeight(height: 100)
    print("놀이기구 타는 것 가능: \(isChecked)")    
} catch HeightError.maxHeight, HeightError.minHeight {   // 케이스 나열도 가능해짐  
    print("놀이기구 타는 것 불가능")  
}
```